### PR TITLE
[move-lang] Update compiler API

### PR DESCRIPTION
**Pull request by tnowacki**
- Overhaul compiler API for a 'Compiler' struct entry point
- Will make it a lot easier to deal with a extra flags

## Motivation

- We will need a bunch of extra flags especially with named addresses. The struct-based entrypoint will make that easier
- I also tried to clean up the pattern of `match result { PassResult(ast) => ast, _ => unreachable!() }`  by using const-generics. If conditions on const generics ever land we will get some neat stuff with `run`. It might be a bit overboard right now but I thought it was better than stamping out a bunch of different compiler-structs for each pass

## Test Plan

- ran the tests

### CODE REVIEW

**Comment by bob-wilson** [comment]
I'm probably not going to have time to give you a detailed review on this, so please don't wait for me.

**Review by tzakian — COMMENTED** [review]
Love the new API. Using const generics is really useful here I think.

One or two small things, but other than that this looks good to me. Feel free to ping when you want to land and I can approve.

File: language/move-lang/src/command_line/compiler.rs [hunk 1]
```diff
@@ -0,0 +1,704 @@
...
+
+pub type Pass = u8;
+pub const EMPTY_COMPILER: Pass = 0;
+pub const PASS_PARSER: Pass = 1;
+pub const PASS_EXPANSION: Pass = 2;
+pub const PASS_NAMING: Pass = 3;
+pub const PASS_TYPING: Pass = 4;
+pub const PASS_HLIR: Pass = 5;
+pub const PASS_CFGIR: Pass = 6;
+pub const PASS_COMPILATION: Pass = 7;
```
[the remark this answers is not in the document]
Comment (tnowacki) [reply language/move-lang/src/command_line/compiler.rs:53]: Unfortunately, rust syntax doesn't allow this with `{}` which is why I went with a bunch of top level ones. I thought `run::<PASS_PARSER>()` read a bit better than `run::<{ pass::PARSER }>()`

File: language/move-lang/src/command_line/compiler.rs [hunk 2]
```diff
@@ -0,0 +1,704 @@
...
+        assert!(P > EMPTY_COMPILER);
+        assert!(self.program.is_some());
+        assert!(self.program.as_ref().unwrap().equivalent_pass() == P);
+        assert!(
+            P <= PASS_COMPILATION,
+            "Invalid pass for run_to. Initial pass is too larget."
+        );
+        assert!(
+            P <= TARGET,
+            "Invalid pass for run_to. Target is greater than current pass"
```
[the remark this answers is not in the document]
Comment (tnowacki) [reply language/move-lang/src/command_line/compiler.rs:181]: ah yeah good catch

File: language/move-lang/src/command_line/compiler.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (tzakian) [reply language/move-lang/src/command_line/compiler.rs:53]: Ahh I see. Definitely agree that `run::<PASS_PARSER>` is better than `run::<{pass::PARSER}>`, so I think keeping it like this then makes sense.

**Review by tzakian — DISMISSED** [review]

**Pull request merged**